ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java

```diff
@@ -428,12 +428,18 @@ Map<String, String> getToolExecuteParams(MLToolSpec toolSpec, Map<String, String
             }
         }
 
+        // Override all parameters in tool config to tool execution parameters as the config contains the static parameters.
+        if (toolSpec.getConfigMap() != null && !toolSpec.getConfigMap().isEmpty()) {
+            executeParams.putAll(toolSpec.getConfigMap());
+        }
+
         if (executeParams.containsKey("input")) {
             String input = executeParams.get("input");
             StringSubstitutor substitutor = new StringSubstitutor(executeParams, "${parameters.", "}");
             input = substitutor.replace(input);
             executeParams.put("input", input);
         }
+
```

Comment:
I'm not sure if we have a chicken-and-egg problem here regarding the 'input' parameter. In #2918 'SearchIndexTool inside conversational agent', there is the need of a placeholder for the last output (i.e. action input) of the former agent step. This output/action input must be inserted into the OpenSearch query defined statically inside the new config, by placeholder.

There exists two possibilities for this:
1. There is an existing placeholder reference to the last output/action input that can be used, i.e. a dedicated parameter list entry. I'm not aware of such an entry, but if such a placeholder exists, the current solution would fix this scenario with a static config of e.g. ["input='"query": {"match": {"body": {"query": "${someUnknownPlaceholderToActionInput}"}}'"].
2. There is no such a placeholder, the last output/action input is just the 'input' entry in the current 'params' method parameter - so the placeholder inside the static config would be 'input'. This value inside 'params' is currently overwritten with 'input' from the static config, **before** substitution. So it is lost, right?

The final behavior should be as follows:
* method params: "input='What is Frodos age?'"  (comes from former agent step)
* static config params "input='"query": {"match": {"body": {"query": "${parameters.input}"}}'"

=> final params for tool execution should be then
"input='"query": {"match": {"body": {"query": "What is Frodos age?"}}'"

This could be achieved as follows:
```java
       // remember former input
       String formerInput = executeParams.get("input");

        if (toolSpec.getConfigMap() != null && !toolSpec.getConfigMap().isEmpty()) {
            executeParams.putAll(toolSpec.getConfigMap());
        }

        if (executeParams.containsKey("input")) {
            String input = executeParams.get("input");

            // set former input as substitution value
           if(formerInput != null)
               executeParams.put("input", formerInput); // add it as "formerInput" or "actionInput" to introduce a new placeholder

            StringSubstitutor substitutor = new StringSubstitutor(executeParams, "${parameters.", "}");
            input = substitutor.replace(input);
            executeParams.put("input", input);
```

The lack is that you can not reference to the input specified in the new static config anymore, which is maybe wanted behavior. To overcome this, we could introduce a new placeholder for the former input (aka the placeholder I mentioned in 1.). This would be a new parameter list entry, e.g. 'formerInput' (just add the formerInput with `executeParams.put("formerInput", formerInput);`).

The static config could be defined now as ["input='"query": {"match": {"body": {"query": "${parameters.formerInput}"}}'"].

Another name for 'formerInput' could be 'actionInput', maybe this is more intuitive.


The same is true for `MLFlowAgentRunner` of course, and for `AgentUtils`, whereby in AgentUtils the former input is defined in the method parameter 'actionInput'. So the only change here would be `executeParams.put("formerInput", actionInput);` or `executeParams.put("actionInput", actionInput);` respectively.

Reply:
Thanks @reuschling for the detailed comment. Actually we are the first possibility you mentioned.
#### Flow Agent
We have a dedicated parameter parameter for the previous tool output here https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLFlowAgentRunner.java#L112. So you can reference it in the current tool input parameter as ${parameters.previous_tool_name.output}.
#### Conversational Agent
We take the action input for tools only from LLM response, i.e, the agent totally depends on LLM to do the ReAct. User can define a tool input to override the action input, but I don't see a use case for a placeholder of action input.

Reply:
Good to hear that there is already a parameter for flow agent.
For Conversational agent, my example is exactly my use case. I want to predefine an Opensearch query, and the tool description says what's to insert. Thus, the previous llm step (the agent) doesn't have to generate the opensearch syntax, only a simple drop in, so I can guarantee this query will be executed, under all circumstances.
More specifically, I want to e. g. count documents with a query string inside.
Without a placeholder, the agent would have no possibility to adjust the query - only static queries independent from the previous steps in the thought chain would be possible.

Reply:
@jngz-es any comment to my use case? This is my origin need and the reason behind #2918

Reply:
Got your point. I created another issue to track it. https://github.com/opensearch-project/ml-commons/issues/3134

Reply:
Is there a chance left that it make it to 2.18?

Reply:
@reuschling The backport PR has been merged into the 2.x branch. The 2.18 branch will include this PR.

Reply:
I mean the conversational agent placeholder now tracked in #3134

Reply:
Hi @reuschling , I raised a pr https://github.com/opensearch-project/ml-commons/pull/3200, could you take a look?